### DISCUSSION TITLE
Option to disable tree-shaking / dead-code-elimination on SWC bundler

### DISCUSSION THREAD

**Discussion by aalykiot** [swc-project/swc#3094] [Q&A]
Hi everyone!

Quick question. Is there any way to disable the tree-shaking the SWC bundler performs ?? 

Here is my use case. I'm trying to build a tiny JavaScript runtime using an engine that does not support es-modules, so my "big" idea for giving modules to the users was to use swc_bundler to bundle everything into a single script (no imports whatsoever) and then pass the generated file to the engine. That works fine until we hit the `REPL problem`.

> No matter where the source code comes from (files or the repl) it always passes through the bundler.

In the REPL I want to support imports as well:

```js
>> import confetti from 'https://cdn.skypack.dev/canvas-confetti';
undefined
>> confetti();
...
```

Now the problem arises when I pass this line `import confetti from 'https://cdn.skypack.dev/canvas-confetti';` through the bundler. Since we're not using the import immediately the bundler sees it as an unused import (or something) and eliminates it from the final bundle returning just `\n` as the output.

> Another interesting result is when we run `>> 1 + 1` through the REPL (same problem).

So, on the rust side is there a way to disable the bundler's optimizations ??

I'm guessing this is where the optimization happens https://github.com/swc-project/swc/blob/main/crates/swc_bundler/src/bundler/optimize.rs (i might be wrong though 🤷‍♂️)

```rs
pub(super) fn optimize(&self, mut node: Module) -> Module {
  self.run(|| {
      if !self.config.disable_inliner {
          node = node.fold_with(&mut constant_propagation())
      }

      node = node.fold_with(&mut Repeat::new(dce::dce(dce::Config {            <--------- right here.
          // TODO(kdy1): Apply mark to wrapped esms and use it at here.
          module_mark: None,
      })));

      node
  })
}
```

If there is no available solution at the moment, is adding another option inside the config a possibility ??

```rs
pub struct Config {
    pub require: bool,
    pub disable_inliner: bool,
    pub disable_hygiene: bool,
    pub disable_fixer: bool,
    pub disbale_optimazations: bool,    <---------
    pub external_modules: Vec<JsWord>,
    pub module: ModuleType,
}
```

```rs
pub(super) fn optimize(&self, mut node: Module) -> Module {
  self.run(|| {
      if !self.config.disable_inliner {
          node = node.fold_with(&mut constant_propagation())
      }
      if !self.config.disbale_optimazations {       <---------
          node = node.fold_with(&mut Repeat::new(dce::dce(dce::Config {
              // TODO(kdy1): Apply mark to wrapped esms and use it at here.
              module_mark: None,
          })));
      }
      node
  })
}
```

I know my use case is quite rare trying to make a bundler less performant 😅 , but is there any solution I'm currently missing ??

Thanks! 🙂

**Comment by kwonoj** [comment]
I do not think there is out of the box configuration you can control those behaviors at the moment, or in the future. There is actually some discussions around tree shaking controllability but that doesn't gaurantee anything if there'll be public interface or not.

I can't comment around specified location yet, but I agree for now trying to modifying code paths in rust modules seems fastest workaround you can try.

**Comment by kdy1** [comment]
If you are not going to use imported identifier, you can remove import specifier to indicate swc that it's an import for side-effects.

**Reply by kdy1** [reply]
Also, I'm fine with a PR to add an option for disabling tree shaking.

**Reply by aalykiot** [reply]
I think this works only if you import a file with side effects and no exports at all (the unused exported stuff will be dropped by the bundler).

**Reply by aalykiot** [reply]
Adding a tree-shaking disable option seems fairly straightforward so I'll try to hack something today and open a PR (the code above seems to do the trick).
